operator/v1alpha1/types_kueue.go

```diff
@@ -0,0 +1,277 @@
+	// The optional prefix and the name are separate by a forward slash (/).
+	// +kubebuilder:validation:MaxLength=317
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:XValidation:rule="!format.qualifiedName().validate(self).hasValue()",message="a qualified name must consist of a lower-case alphanumeric and hyphenated string of at most 63 characters in length, starting and ending with alphanumeric chracters. The name may be optionally prefixed with a subdomain consisting of lower-case alphanumeric characters, hyphens and periods, of at most 253 characters in length. Each period separated segment within the subdomain must start and end with an alphanumeric character."
+	// +optional
+	Key string `json:"key,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=ByWorkload;Disabled
+type KueueGangSchedulingPolicyOptions string
+
+const (
+	KueueGangSchedulingPolicyEvictNotReadyWorkloads KueueGangSchedulingPolicyOptions = "ByWorkload"
+	KueueGangSchedulingPolicyDisabled               KueueGangSchedulingPolicyOptions = "Disabled"
+)
+
+// +kubebuilder:validation:Enum=Parallel;Sequential
+type KueueGangSchedulingAdmissionOptions string
+
+const (
+	KueueGangSchedulingAdmissionOptionsSequential KueueGangSchedulingAdmissionOptions = "Sequential"
+	KueueGangSchedulingAdmissionOptionsParallel   KueueGangSchedulingAdmissionOptions = "Parallel"
+)
+
+// Kueue provides the ability to admit workloads all in one (gang admission)
+// and evicts workloads if they are not ready within a specific time.
+type KueueGangSchedulingPolicy struct {
+	// policy allows for changing the kinds of gang scheduling Kueue does.
+	// This is an optional field.
+	// The allowed values are ByWorkload and Disabled.
+	// The default value will be Disabled.
+	// ByWorkload allows for configuration how admission is performed
+	// for Kueue.
+	// +optional
+	Policy KueueGangSchedulingPolicyOptions `json:"policy"`
+	// byWorkload controls how admission is done.
+	// When admission is set to Sequential, only pods from the currently processing workload will be admitted.
+	// Once all pods from the current workload are admitted, and ready, Kueue will process the next workload.
+	// Sequential processing may slow down admission when the cluster has sufficient capacity for multiple workloads,
+	// but provides a higher guarantee of workloads scheduling all pods together successfully.
+	// When set to Parallel, pods from any workload will be admitted at any time.
+	// This may lead to a deadlock where workloads are in contention for cluster capacity and
+	// pods from another workload having successfully scheduled prevent pods from the current workload scheduling.
+	// +kubebuilder:validation:XValidation:rule="self.policy==ByWorkload",message="byWorkload is only valid if policy equals ByWorkload"
+	// +optional
+	ByWorkload KueueGangSchedulingAdmissionOptions `json:"byWorkload"`
```

Comment:
Will we only ever have a static admission option of `Sequential` or `Parallel` here? With the Kueue configuration being in beta would it be worth having a proper struct here with a nested `admissionPolicy` field (or something like that)?

Reply:
It only depends on Policy of ByWorkload.

So if Policy is set to ByWorkload, then we can either block admission or allow admission in parallel.

Technically if gang scheduling is disabled, all admission is parallel but there are no evictions of workloads.

Reply:
Just because someone hasn't asked for this _yet_ doesn't mean we should design the API in a way that prohibits us from doing this in the future. I may need to read up a bit on more recent conversations, but what does moving this to the workload level mean?

Reply:
I am in agreement with Joel here. I think the benefit of making this a struct, and therefore more flexible for future expansion, far outweighs the potential downside of no one asking us for this and having to do one more layer of nesting to set what they want to set

Reply:
So for workloads level, it would mean that a user creating a batch job would be able to specify overrides for these values. It would no longer be necessary to put this on the config level.

And a user could set these up on the workload that they are submitted rather than relying on a global config.

Reply:
Why could we just add new fields and require that they are set for ByWorkload and add these to the struct we have.

Why define them all now in case there is a need for it someday? We don't have any explicit customer ask for any of these values so I think leaving it alone for now is sufficient. We can always add a new field to an API. If you want a struct, I can give you an empty struct that we can add to in the future but I think its distracting to try and define these APIs where we have no use case for them yet.

If we need to add new fields, there are in a struct so we can add new features. I don't expect `WaitForPodsReady` Kueue API to be stable anytime soon so I don't want to expose every field in the Kueueu config API in case there is a need for it.

Reply:
You can't add new required fields in the future. That is a breaking change.

The point of using a struct is that you have a valid way to expand the options in the future without introducing potentially breaking changes or patterns that don't align with ones that are typically used in OpenShift's API.

It is a bit of an anti-pattern to have a discriminated union that results in you needing to set multiple fields related to the "type".

Reply:
All these fields are optional. We would not change the tags optional but we enforce via validation that these values can be set if ByWorkload is set.

So are you saying something like:

```
GangScheduling:
  policy: ByWorkload
  byWorkload:
    admission: Parallel
```

We can always add the extra fields for this if there is a request for it but we at least have the struct structure?

Reply:
Bryce and I met offline. I pushed a change to add a byWorkload nested field but I only included admission.

Reply:
resolving this.